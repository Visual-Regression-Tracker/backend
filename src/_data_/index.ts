import { Baseline, Build, ImageComparison, Project, TestRun, TestVariation } from '@prisma/client';
import { PNG } from 'pngjs';


export const TEST_PROJECT: Project = {
  id: '1',
  name: 'Test Project',
  buildsCounter: 2,
  maxBuildAllowed: 100,
  maxBranchLifetime: 30,
  mainBranchName: 'master',
  createdAt: new Date(),
  updatedAt: new Date(),
  autoApproveFeature: true,
  imageComparisonConfig: '{ "threshold": 0.1, "ignoreAntialiasing": true, "allowDiffDimensions": false }',
  imageComparison: ImageComparison.pixelmatch,
};

export const TEST_BUILD: Build = {
  id: 'a9385fc1-884d-4f9f-915e-40da0e7773d5',
  ciBuildId: 'ciBuildId',
  number: 2345,
  branchName: 'develop',
  status: 'new',
  projectId: 'e0a37894-6f29-478d-b13e-6182fecc715e',
  updatedAt: new Date(),
  createdAt: new Date(),
  userId: '2341235',
  isRunning: true,
};

export const generateBaseline = (baseline?: Partial<Baseline>): Baseline => {
  return {
    id: 'baselineId',
    baselineName: 'baselineName',
    testVariationId: 'testVariationId',
    testRunId: 'testRunId',
    userId: 'userId',
    updatedAt: new Date(),
    createdAt: new Date(),
    ...baseline,
  };
};

type TestVariationWithBaselines = TestVariation & {
  baselines: Baseline[];
}

export const generateTestVariation = (
  testVariation?: Partial<TestVariation>,
  baselines?: Baseline[]
): TestVariationWithBaselines => {
  return {
    id: '123',
    projectId: 'project Id',
    name: 'Test name',
    baselineName: 'baselineName',
    os: 'OS',
    browser: 'browser',
    viewport: 'viewport',
    device: 'device',
    customTags: '',
    ignoreAreas: '[]',
    comment: 'some comment',
    createdAt: new Date(),
    updatedAt: new Date(),
    branchName: 'master',
    ...testVariation,
    baselines: baselines ?? [generateBaseline()],
  };
};

export const generateTestRun = (testRun?: Partial<TestRun>): TestRun => {
  return {
    id: '10fb5e02-64e0-4cf5-9f17-c00ab3c96658',
    imageName: '1592423768112.screenshot.png',
    diffName: 'diffName',
    diffPercent: 12,
    diffTollerancePercent: 1,
    pixelMisMatchCount: 123,
    status: 'new',
    buildId: '146e7a8d-89f0-4565-aa2c-e61efabb0afd',
    testVariationId: '3bc4a5bc-006e-4d43-8e4e-eaa132627fca',
    projectId: '3bc4a5bc-006e-4d43-8e4e-eaa132627fcc',
    updatedAt: new Date(),
    createdAt: new Date(),
    name: 'ss2f77',
    browser: 'chromium',
    device: null,
    os: null,
    viewport: '1800x1600',
    customTags: '',
    baselineName: null,
    ignoreAreas: '[]',
    tempIgnoreAreas: '[]',
    comment: 'some comment',
    baselineBranchName: 'master',
    branchName: 'develop',
    merge: false,
    ...testRun,
  };
};

export const generatePng = (width: number, height: number): PNG => {
  const png = new PNG({ width, height });
  for (let y = 0; y < png.height; y++) {
    for (let x = 0; x < png.width; x++) {
      const idx = (png.width * y + x) << 2;
      png.data[idx] = 255; // red
      png.data[idx + 1] = 255; // green
      png.data[idx + 2] = 255; // blue
      png.data[idx + 3] = 255; // alpha (0 is transparent)
    }
  }
  return png.pack();
};
