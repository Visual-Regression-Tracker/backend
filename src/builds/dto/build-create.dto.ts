import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { IsString, IsOptional, IsNotEmpty } from 'class-validator';

export class CreateBuildDto {
  @ApiPropertyOptional()
  @IsOptional()
  @IsString()
  @IsNotEmpty()
  readonly ciBuildId?: string;

  @ApiPropertyOptional()
  @IsOptional()
  @IsString()
  @IsNotEmpty()
  readonly branchName?: string;

  @ApiProperty()
  @IsString()
  readonly project: string;
}
