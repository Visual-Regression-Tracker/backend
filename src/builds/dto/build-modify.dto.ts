export interface ModifyBuildDto {
  ciBuildId?: string;
  isRunning?: boolean;
}
