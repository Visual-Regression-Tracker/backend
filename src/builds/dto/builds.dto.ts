import { ApiProperty } from '@nestjs/swagger';
import { Build } from '../build.entity';

export class BuildDto {
  @ApiProperty()
  readonly id: number;

  @ApiProperty()
  readonly projectName: string;

  @ApiProperty()
  readonly branchName: string;

  @ApiProperty()
  readonly status: string;

  @ApiProperty()
  readonly createdBy: string;

  @ApiProperty()
  readonly createdAt: Date;

  constructor(build: Build) {
    this.id = build.id;
    this.createdAt = build.createdAt;
    this.status = build.status;
    this.branchName = build.branchName;
    this.createdBy = build.user.email;
    this.projectName = build.project.name;
  }
}
