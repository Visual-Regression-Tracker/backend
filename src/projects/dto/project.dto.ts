import { ApiProperty } from '@nestjs/swagger';
import { IsUUID, IsString, IsNumber, IsBoolean, IsEnum, IsJSON, IsDate } from 'class-validator';
import { ImageComparison, Project } from '@prisma/client';
import { Type } from 'class-transformer';

export class ProjectDto implements Project {
  @ApiProperty()
  @IsUUID()
  readonly id: string;

  @ApiProperty()
  @IsNumber()
  readonly buildsCounter: number;

  @ApiProperty()
  @IsString()
  readonly name: string;

  @ApiProperty()
  @IsString()
  readonly mainBranchName: string;

  @ApiProperty()
  @IsDate()
  @Type(() => Date)
  readonly createdAt: Date;

  @ApiProperty()
  @IsDate()
  @Type(() => Date)
  readonly updatedAt: Date;

  @ApiProperty()
  @IsBoolean()
  autoApproveFeature: boolean;

  @ApiProperty()
  @IsEnum(ImageComparison)
  imageComparison: ImageComparison;

  @ApiProperty()
  @IsNumber()
  maxBuildAllowed: number;

  @ApiProperty()
  @IsNumber()
  maxBranchLifetime: number;

  @ApiProperty()
  @IsJSON()
  imageComparisonConfig: string;
}
