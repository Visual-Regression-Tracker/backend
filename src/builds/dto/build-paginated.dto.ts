import { ApiProperty } from '@nestjs/swagger';
import { Pagination } from '../../shared/dto/pagination.dto';
import { BuildDto } from './build.dto';

export class PaginatedBuildDto extends Pagination {
  @ApiProperty({ type: BuildDto, isArray: true })
  data: BuildDto[];
}
