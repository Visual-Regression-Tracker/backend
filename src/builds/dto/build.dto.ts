import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Build, TestStatus } from '@prisma/client';
import { TestRunDto } from 'src/test-runs/dto/testRun.dto';

export class BuildDto {
  @ApiProperty()
  id: string;

  @ApiPropertyOptional()
  ciBuildId?: string;

  @ApiProperty()
  number: number;

  @ApiProperty()
  branchName: string;

  @ApiProperty()
  status: string;

  @ApiProperty()
  projectId: string;

  @ApiProperty()
  updatedAt: Date;

  @ApiProperty()
  createdAt: Date;

  @ApiPropertyOptional()
  userId?: string;

  @ApiProperty()
  passedCount: number;
  @ApiProperty()
  unresolvedCount: number;
  @ApiProperty()
  failedCount: number;
  @ApiProperty()
  isRunning: boolean;
  @ApiProperty()
  merge: boolean;

  constructor(build: Build & { testRuns?: TestRunDto[] }) {
    this.id = build.id;
    this.ciBuildId = build.ciBuildId;
    this.number = build.number;
    this.branchName = build.branchName;
    this.userId = build.userId;
    this.status = build.status;
    this.projectId = build.projectId;
    this.updatedAt = build.updatedAt;
    this.createdAt = build.createdAt;
    this.isRunning = build.isRunning;

    this.passedCount = 0;
    this.unresolvedCount = 0;
    this.failedCount = 0;

    this.merge = false;
    if (build.testRuns) {
      // determine if merge
      this.merge = build.testRuns.some((testRun) => testRun.merge);

      // calculate statistics
      build.testRuns.forEach((testRun) => {
        switch (testRun.status) {
          case TestStatus.autoApproved:
          case TestStatus.approved:
          case TestStatus.ok: {
            this.passedCount += 1;
            break;
          }
          case TestStatus.unresolved:
          case TestStatus.new: {
            this.unresolvedCount += 1;
            break;
          }
          case TestStatus.failed: {
            this.failedCount += 1;
            break;
          }
        }
      });
    }

    if (!build.testRuns || build.testRuns.length === 0) {
      this.status = 'new';
    } else {
      this.status = 'passed';
    }
    if (this.failedCount > 0) {
      this.status = 'failed';
    }
    if (this.unresolvedCount > 0) {
      this.status = 'unresolved';
    }
  }
}
