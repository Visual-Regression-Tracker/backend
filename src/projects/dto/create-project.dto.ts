import { OmitType } from '@nestjs/swagger';
import { UpdateProjectDto } from './update-project.dto';

export class CreateProjectDto extends OmitType(UpdateProjectDto, ['id']) {}
