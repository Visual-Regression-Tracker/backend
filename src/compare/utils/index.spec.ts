import { PNG } from 'pngjs';
import { IgnoreAreaDto } from '../../test-runs/dto/ignore-area.dto';
import { generatePng } from '../../_data_';
import { applyIgnoreAreas } from '.';

describe('utils', () => {
  it.each<[PNG, IgnoreAreaDto[]]>([
    [generatePng(3, 3), []],
    [generatePng(3, 3), [{ x: 0, y: 0, width: 3, height: 3 }]],
    [generatePng(3, 3), [{ x: -3, y: -3, width: 3, height: 3 }]],
    [generatePng(3, 3), [{ x: 1, y: 1, width: 5, height: 5 }]],
  ])('applyIgnoreAreas', (image, ignoreAreas) => {
    expect(applyIgnoreAreas(image, ignoreAreas).data).toMatchSnapshot();
  });
});
