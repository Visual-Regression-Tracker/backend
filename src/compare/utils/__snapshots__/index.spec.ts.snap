// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils applyIgnoreAreas 1`] = `
{
  "data": [
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`utils applyIgnoreAreas 2`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`utils applyIgnoreAreas 3`] = `
{
  "data": [
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`utils applyIgnoreAreas 4`] = `
{
  "data": [
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    255,
    255,
    255,
    255,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;
